lib/baby_squeel/join_dependency.rb

```diff
@@ -35,43 +35,27 @@ def initialize(relation)
       # a list (in order of chaining) of associations and finding
       # the respective JoinAssociation at each level.
       def find_alias(associations)
-        table = find_join_association(associations).table
-        reconstruct_with_type_caster(table, associations)
+        # If we tell join_dependency to construct its tables, Active Record
+        # handles building the correct aliases and attaching them to its
+        # JoinDepenencies.
+        join_dependency.send(:construct_tables!, join_dependency.send(:join_root))
+        join_association = find_join_association(associations)
+
+        join_association.table
       end
 
       private
 
       def find_join_association(associations)
-        associations.inject(join_dependency.send(:join_root)) do |parent, assoc|
-          parent.children.find do |join_association|
-            reflections_equal?(
-              assoc._reflection,
-              join_association.reflection
-            )
-          end
+        current = join_dependency.send(:join_root)
+        
+        associations.each do |association|
+          name = association._reflection.name
+          current = current.children.find { |c| c.reflection.name == name }
+          break if current.nil?
```

Comment:
Could you tell me more about the rationale for the `break` here? 

If the association isn't found, that would mean it doesn't exist, which should be an error, right?. The way this is written, it's saying that if an association isn't found, just use its parent, which I don't think is the behavior we want.

Reply:
No, if `current` becomes `nil`, then the next iteration of `associations` will produce a `NoMethodError` for `#children`.  This breaks out of the `#each`, allowing `nil` to be returned from the method.

Reply:
Ahh, gotcha. Okay, the logic makes sense. In what scenario would we be okay with `current` not being found? If we don't find the association at a given path, shouldn't that be an error?

Reply:
It might only be possible if the Active Record implementation changes. But I'm not sure if we want `#find_join_association` to return `nil` or raise an exception in that case. The way the code is now, a returned `nil` from here will result in `NoMethodError: undefined method 
'table' for nil:NilClass` pointing to the line `join_association.table` (line 44). But if we ever needed to find a join association as part of resolving a name in the DSL, returning `nil` here might be nicer than raising an exception.

Reply:
I guess I was going off `Enumerable#find` semantics, where it returns `nil` if the block never returns truthy.